File: engine/app/contract.py
"""The Rust ↔ Python contract.

These models are the Pydantic mirror of `orchestrator/src/contract.rs` (serde)
and the JSON shapes documented in `README.md`. All three must stay in sync.

Wire format is snake_case (field names as-is) with enums serialized to their
lowercase string value.
"""

from __future__ import annotations

from enum import Enum
from typing import Any, Optional

from pydantic import BaseModel

# ---------------------------------------------------------------------------
# Enums  (str-mixin so they serialize to their lowercase string value)
# ---------------------------------------------------------------------------


class Framework(str, Enum):
    langgraph = "langgraph"
    openai = "openai"
    claude = "claude"


class Category(str, Enum):
    injection = "injection"
    tool_misuse = "tool_misuse"
    goal_hijack = "goal_hijack"
    exfiltration = "exfiltration"
    edge_case = "edge_case"


class Severity(str, Enum):
    low = "low"
    medium = "medium"
    high = "high"


class DetectionMethod(str, Enum):
    tool_call = "tool_call"
    output_contains = "output_contains"
    output_absent = "output_absent"
    error_or_crash = "error_or_crash"
    judge_only = "judge_only"


# ---------------------------------------------------------------------------
# Core shapes
# ---------------------------------------------------------------------------


class ToolSpec(BaseModel):
    name: str
    description: str
    params_schema: dict[str, Any]


class AgentSpec(BaseModel):
    framework: Framework
    entry_file: str
    agent_variable_name: str
    tools: list[ToolSpec]
    system_prompt: Optional[str]


class IntrospectResult(BaseModel):
    """A repo may expose more than one agent/graph (e.g. a retrieval template
    with a separate indexer graph), each with its own tools and prompt. The
    introspection result is therefore a list. Single-agent repos return a
    one-element list. An empty list is never returned — the engine raises
    AgentNotFound (HTTP 422) instead."""

    agents: list[AgentSpec]


class Detection(BaseModel):
    """How a scorer should detect this test's failure from the agent's runtime
    output. `tool_name`/`arg_name`/`arg_pattern`/`needle` use "" when not
    applicable to the chosen method. Transient generate->score data — NOT
    persisted to test_results."""

    method: DetectionMethod
    tool_name: str = ""
    arg_name: str = ""
    arg_pattern: str = ""  # regex; validated to compile at generation time
    needle: str = ""
    rationale: str


class TestCase(BaseModel):
    id: str
    category: Category
    prompt: str
    expected_failure_mode: str
    severity: Severity
    detection: Detection


class RunResult(BaseModel):
    test_case_id: str
    agent_output: str
    tool_calls: list[dict[str, Any]]
    passed: bool
    failure_reason: Optional[str]


# ---------------------------------------------------------------------------
# Engine request/response envelopes
# ---------------------------------------------------------------------------


class IntrospectRequest(BaseModel):
    repo_path: str
    framework: Framework


class GenerateRequest(BaseModel):
    agent_spec: AgentSpec


class GenerateResponse(BaseModel):
    test_cases: list[TestCase]
    # Actual number of LLM API calls made while generating these test cases.
    # The orchestrator reconciles its run-level budget against this real count
    # instead of a fixed estimate. 0 when generated offline (no API key).
    llm_calls: int = 0


class ScoreRequest(BaseModel):
    test_cases: list[TestCase]
    results: list[RunResult]


class ScoreResponse(BaseModel):
    scored: list[RunResult]
    overall_passed: bool
    summary: str
    # Actual number of LLM API calls made while judging (0 while /score is
    # stubbed; ready for when scoring becomes a real LLM judge).
    llm_calls: int = 0

File: orchestrator/src/contract.rs
//! The Rust ↔ Python contract.
//!
//! These types are the serde mirror of `engine/app/contract.py` (Pydantic) and
//! the JSON shapes documented in `README.md`. All three must stay in sync.
//!
//! Wire format is `snake_case` (field names as-is) with enums serialized to
//! their lowercase string value.

use serde::{Deserialize, Serialize};

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Framework {
    Langgraph,
    Openai,
    Claude,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Category {
    Injection,
    ToolMisuse,
    GoalHijack,
    Exfiltration,
    EdgeCase,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Severity {
    Low,
    Medium,
    High,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DetectionMethod {
    ToolCall,
    OutputContains,
    OutputAbsent,
    ErrorOrCrash,
    JudgeOnly,
}

// ---------------------------------------------------------------------------
// Core shapes
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolSpec {
    pub name: String,
    pub description: String,
    /// Free-form JSON Schema object.
    pub params_schema: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentSpec {
    pub framework: Framework,
    pub entry_file: String,
    pub agent_variable_name: String,
    pub tools: Vec<ToolSpec>,
    pub system_prompt: Option<String>,
}

/// A repo may expose more than one agent/graph, each with its own tools and
/// prompt. Introspection therefore returns a list; single-agent repos yield a
/// one-element list. The list is never empty — the engine returns 422
/// (AgentNotFound) instead.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IntrospectResult {
    pub agents: Vec<AgentSpec>,
}

/// How a scorer detects this test's failure from the agent's runtime output.
/// Empty-string fields are "not applicable" for the chosen method. Transient
/// generate->score data — NOT persisted to `test_results`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Detection {
    pub method: DetectionMethod,
    #[serde(default)]
    pub tool_name: String,
    #[serde(default)]
    pub arg_name: String,
    #[serde(default)]
    pub arg_pattern: String,
    #[serde(default)]
    pub needle: String,
    pub rationale: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TestCase {
    pub id: String,
    pub category: Category,
    pub prompt: String,
    pub expected_failure_mode: String,
    pub severity: Severity,
    pub detection: Detection,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RunResult {
    pub test_case_id: String,
    pub agent_output: String,
    /// Array of free-form tool-call objects.
    pub tool_calls: Vec<serde_json::Value>,
    pub passed: bool,
    pub failure_reason: Option<String>,
}

// ---------------------------------------------------------------------------
// Engine request/response envelopes
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IntrospectRequest {
    pub repo_path: String,
    pub framework: Framework,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GenerateRequest {
    pub agent_spec: AgentSpec,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GenerateResponse {
    pub test_cases: Vec<TestCase>,
    /// Actual number of LLM API calls made while generating these test cases.
    /// The orchestrator reconciles its run-level budget against this real count
    /// instead of a fixed estimate. 0 when generated offline (no API key).
    #[serde(default)]
    pub llm_calls: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScoreRequest {
    pub test_cases: Vec<TestCase>,
    pub results: Vec<RunResult>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScoreResponse {
    pub scored: Vec<RunResult>,
    pub overall_passed: bool,
    pub summary: String,
    /// Actual number of LLM API calls made while judging (0 while /score is
    /// stubbed; ready for when scoring becomes a real LLM judge).
    #[serde(default)]
    pub llm_calls: u32,
}
